test/performance/latency-controller/internal/controller/controller.go

```diff
@@ -0,0 +1,737 @@
+	totalE2EHistogram          *prometheus.HistogramVec
+	workloadCountGauge         *prometheus.GaugeVec
+}
+
+//+kubebuilder:rbac:groups=*,resources=*,verbs=get;list;watch
+//+kubebuilder:rbac:groups=control.kubestellar.io,resources=bindingpolicies;workstatuses,verbs=get;list
+//+kubebuilder:rbac:groups=work.open-cluster-management.io,resources=manifestworks;appliedmanifestworks,verbs=get;list
+
+func (r *GenericLatencyCollectorReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	r.cache = make(map[string]*PerWorkloadCache)
+
+	// Precompute GVK to GVR mappings
+	mapper := mgr.GetRESTMapper()
+	r.gvkToGVR = make(map[schema.GroupVersionKind]schema.GroupVersionResource)
+	for _, gvk := range r.DiscoveredResources {
+		mapping, err := mapper.RESTMapping(gvk.GroupKind(), gvk.Version)
+		if err == nil {
+			r.gvkToGVR[gvk] = mapping.Resource
+		}
+	}
+
+	// Create a controller builder with proper unstructured handling
+	preds := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			if obj, ok := e.Object.(*unstructured.Unstructured); ok {
+				gvk := obj.GetObjectKind().GroupVersionKind()
+				if gvk.Kind == "" {
+					for _, discoveredGVK := range r.DiscoveredResources {
+						if discoveredGVK.Kind == obj.GetKind() {
+							obj.SetGroupVersionKind(discoveredGVK)
+							break
+						}
+					}
+				}
+				return obj.GetObjectKind().GroupVersionKind().Kind != ""
+			}
+			return e.Object.GetObjectKind().GroupVersionKind().Kind != ""
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			if obj, ok := e.ObjectNew.(*unstructured.Unstructured); ok {
+				gvk := obj.GetObjectKind().GroupVersionKind()
+				if gvk.Kind == "" {
+					for _, discoveredGVK := range r.DiscoveredResources {
+						if discoveredGVK.Kind == obj.GetKind() {
+							obj.SetGroupVersionKind(discoveredGVK)
+							break
+						}
+					}
+				}
+				return obj.GetObjectKind().GroupVersionKind().Kind != ""
+			}
+			return e.ObjectNew.GetObjectKind().GroupVersionKind().Kind != ""
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return false
+		},
+		GenericFunc: func(e event.GenericEvent) bool {
+			return false
+		},
+	}
+
+	excludedGroups := map[string]bool{
+		"flowcontrol.apiserver.k8s.io": true,
+		"discovery.k8s.io":             true,
+		"apiregistration.k8s.io":       true,
+		"coordination.k8s.io":          true,
+		"control.kubestellar.io":       true,
+	}
+	excludedResourceNames := map[string]bool{
+		"events":               true,
+		"nodes":                true,
+		"csistoragecapacities": true,
+		"csinodes":             true,
+		"endpoints":            true,
+		"workstatuses":         true,
+	}
+
+	controllerBuilder := ctrl.NewControllerManagedBy(mgr).
+		Named("generic-latency-collector").
+		WithEventFilter(preds)
+
+	for _, gvk := range r.DiscoveredResources {
+		// Skip entire group
+		if excludedGroups[gvk.Group] {
+			continue
+		}
+		// Skip resource name
+		kindLower := strings.ToLower(gvk.Kind)
```

Comment:
This won't work as expected. Because `gvk.Kind` is a singular noun, but `excludedResourceNames` are keyed by plural nouns.

https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md talked about this in its opening paragraphs.